ENH: Caching, advanced mapping and separating events for MISP Feed output bot (#2509)
certtools/intelmq: docs/dev/bot-development.md

- `redis_cache_ttl: int = 15`
- `redis_cache_password: Optional[str] = None`

and provides the methods to cache key-value pairs:

- `cache_exists`
- `cache_get`
- `cache_set`
- `cache_flush`
- `cache_get_redis_instance`

and following methods to cache objects in a queue:

- `cache_lpush`
- `cache_rpop`
- `cache_llen`.

Caching key-value pairs and queue caching are two separated mechanisms. The first is designed
 for arbitrary values, the second one is focused on temporary storing messages (but can handle other

>  data). You won't see caches from one in the another. For example, if adding a key-value pair using

Reviewer: ```suggestion
Caching key-value pairs and queue caching are two different mechanisms. The functions in the first list are designed
 for arbitrary values, while the latter ones are primarily for temporarily storing messages, but can also handle other data
 types. You won't see caches from one in the other. For example, if adding a key-value pair using
```

 `cache_set`, it does not change the value from `cache_llen`, and if adding an element using
 `cache_lpush` you cannot use `check_exists` to look for it.

When using queue-based caching, you have to serialize object to a format accepted by Redis/Valkey
as the underlying storage. For example, to store a message in a queue using bot ID as key, you can
use code like:

```python
msmessage = self.receive_message().to_dict(jsondict_as_string=True)

> self.cache_lpush(self.bot_id, json.dumps(message))

Reviewer: Why not use `message.to_json` directly?
```suggestion
self.cache_lpush(self.bot_id, self.receive_message().to_json(jsondict_as_string=True))
```

Reviewer: Or: `Message.serialize`

Author: Using the first. I have copied the code from the bot, where the JSON representation is used later.

```

and to retrieve a message from the cache:

```python
data = self.cache_pop()
if data is None:
    return # handle empty cache

> message = json.loads(data)
> # to use it as Message object
> message_obj = MessageFactory.from_dict(
>     message, harmonization=self.harmonization, default_type="Event"
> )

Reviewer: Untested:

```suggestion
message_obj = MessageFactory.unserialize(
    data, harmonization=self.harmonization, default_type="Event"
)
```
I think the two optional parameters are unnecessary too.

Author: I'm almost sure the default_type was necessary, but I'll check it again.

Author: The first is recommended, the second is necessary to match the dumping code.

1) if the harmonization is not provided, the code will _read it from the file_, and not use any already loaded version. Reading files and processing JSON is in Python quite expensive operation, it's much better to provide the ready harmonization.
2) the second depends on setting `with_type` when serializing the message. As I didn't set it on serialization, I have to provide type on deserialization. I could do it the other way around, there is no difference - but one or another has to be set.

Reviewer: Ad 1: True, we should probably cache that file -> different topic
Ad 2: `MessageFactory` is the preferred way of convering Messages to JSON back and forth. I agree, that both ways need to be compatible with each other. Let's refactor this doc section later when improving the overall docs.
